File: payment-service/payment-service/Services/PaymentGrpcService.cs
using Grpc.Core;
using payment_service.Entities;
using payment_service.Enums;
using payment_service.Services;
using Payment; // for PaymentService.PaymentServiceBase

namespace payment_service.Services;

public class PaymentGrpcService : PaymentService.PaymentServiceBase
{
    private readonly MyPaymentService _paymentService;
    private readonly ILogger<PaymentGrpcService> _logger;

    public PaymentGrpcService(MyPaymentService paymentService, ILogger<PaymentGrpcService> logger)
    {
        _paymentService = paymentService;
        _logger = logger;
    }

    public override async Task<ProcessPaymentResponse> ProcessPayment(ProcessPaymentRequest request, ServerCallContext context)
    {
        _logger.LogInformation("Processing payment via gRPC for order: {OrderId}", request.OrderId);

        try
        {
            if (string.IsNullOrWhiteSpace(request.Id) || string.IsNullOrWhiteSpace(request.OrderId))
            {
                return new ProcessPaymentResponse
                {
                    PaymentId = request.Id,
                    Status = PaymentStatus.Failed.ToString(),
                    ErrorMessage = "Id and OrderId are required ULIDs from the upstream service."
                };
            }

            var payment = new PaymentEntity
            {
                Id = request.Id,
                OrderId = request.OrderId,
                Amount = new MoneyEntity
                {
                    Amount = request.Amount.Amount,
                    Currency = request.Amount.Currency,
                    Fraction = request.Amount.Fraction
                },
                PaymentMethod = request.PaymentMethod,
                PaymentStatus = Enums.PaymentStatus.Pending
            };

            var result = await _paymentService.ProcessPayment(payment);

            return new ProcessPaymentResponse
            {
                PaymentId = result.Id,
                Status = result.PaymentStatus.ToString(),
                TransactionId = result.Id, // You might want to add a separate transaction ID field
                ClientSecret = result.ClientSecret ?? string.Empty,
                ErrorMessage = result.PaymentStatus == Enums.PaymentStatus.Failed ? "Payment processing failed" : string.Empty
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error processing payment via gRPC for order: {OrderId}", request.OrderId);
            return new ProcessPaymentResponse
            {
                PaymentId = request.Id,
                Status = PaymentStatus.Failed.ToString(),
                ErrorMessage = ex.Message
            };
        }
    }

    public override async Task<GetPaymentStatusResponse> GetPaymentStatus(GetPaymentStatusRequest request, ServerCallContext context)
    {
        _logger.LogInformation("Getting payment status via gRPC for payment: {PaymentId}", request.PaymentId);

        try
        {
            // This would typically query your database for the payment status
            // For now, we'll return a mock response
            return new GetPaymentStatusResponse
            {
                PaymentId = request.PaymentId,
                Status = PaymentStatus.Pending.ToString(),
                OrderId = "mock-order-id",
                Amount = new Payment.Money // <-- Use the generated type here
                {
                    Amount = 1000,
                    Currency = "USD",
                    Fraction = 100
                },
                PaymentMethod = "Card"
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting payment status via gRPC for payment: {PaymentId}", request.PaymentId);
            throw new RpcException(new Status(StatusCode.Internal, ex.Message));
        }
    }

    public override async Task StreamPaymentUpdates(StreamPaymentUpdatesRequest request, IServerStreamWriter<PaymentUpdate> responseStream, ServerCallContext context)
    {
        _logger.LogInformation("Starting payment status stream for payment: {PaymentId}", request.PaymentId);

        try
        {
            // Simulate payment status updates
            var statuses = new[] { "Pending", "Processing", "Completed" };
            
            foreach (var status in statuses)
            {
                if (context.CancellationToken.IsCancellationRequested)
                    break;

                var update = new PaymentUpdate
                {
                    PaymentId = request.PaymentId,
                    Status = status,
                    Message = $"Payment status updated to {status}",
                    Timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds()
                };

                await responseStream.WriteAsync(update);
                await Task.Delay(2000, context.CancellationToken); // Wait 2 seconds between updates
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error streaming payment updates for payment: {PaymentId}", request.PaymentId);
            throw new RpcException(new Status(StatusCode.Internal, ex.Message));
        }
    }
} 

File: old-stuff/core/src/shared/grpc/payment-grpc.service.ts
import { Inject, Injectable, OnModuleInit } from '@nestjs/common';
import { ClientGrpc } from '@nestjs/microservices';
import { firstValueFrom, Observable } from 'rxjs';
import { Money } from 'src/shared/types/money';
import { PaymentMethod } from 'src/checkout/core/entity/payment/payment';

// TypeScript interfaces matching proto definitions
export interface ProcessPaymentResponse {
  payment_id: string;
  status: string;
  transaction_id: string;
  client_secret: string;
  error_message: string;
}

export interface GetPaymentStatusResponse {
  payment_id: string;
  status: string;
  order_id: string;
  amount: { amount: number; currency: string; fraction: number };
  payment_method: string;
}

export interface PaymentUpdate {
  payment_id: string;
  status: string;
  message: string;
  timestamp: number;
}

interface PaymentService {
  processPayment(data: {
    id: string;
    order_id: string;
    amount: { amount: number; currency: string; fraction: number };
    payment_method: string;
  }): Observable<ProcessPaymentResponse>;
  getPaymentStatus(data: {
    payment_id: string;
  }): Observable<GetPaymentStatusResponse>;
  streamPaymentUpdates(data: { payment_id: string }): Observable<PaymentUpdate>;
}

@Injectable()
export class PaymentGrpcService implements OnModuleInit {
  private paymentService: PaymentService;

  constructor(@Inject('PAYMENT_PACKAGE') private client: ClientGrpc) {}

  onModuleInit() {
    this.paymentService =
      this.client.getService<PaymentService>('PaymentService');
  }

  async processPayment(
    id: string,
    orderId: string,
    amount: Money,
    paymentMethod: PaymentMethod,
  ) {
    try {
      const request = {
        id,
        order_id: orderId,
        amount: {
          amount: amount.getAmount(),
          currency: amount.getCurrency(),
          fraction: amount.getFraction(),
        },
        payment_method: paymentMethod,
      };

      const response = await firstValueFrom(
        this.paymentService.processPayment(request),
      );

      return {
        status: 200,
        data: {
          paymentStatus: response.status,
          transactionId: response.transaction_id,
          clientSecret: response.client_secret,
        },
      };
    } catch (error: any) {
      return {
        status: 400,
        data: {
          paymentStatus: 'Failed',
          error: error.message,
        },
      };
    }
  }

  async getPaymentStatus(paymentId: string) {
    try {
      const response = await firstValueFrom(
        this.paymentService.getPaymentStatus({ payment_id: paymentId }),
      );
      return response;
    } catch (error: any) {
      throw new Error(`Failed to get payment status: ${error.message}`);
    }
  }

  streamPaymentUpdates(paymentId: string) {
    return this.paymentService.streamPaymentUpdates({ payment_id: paymentId });
  }
}
